trajopt_ifopt/src/cartesian_position_constraint.cpp

```diff
@@ -81,11 +81,23 @@ void CartPosConstraint::SetBounds(const std::vector<ifopt::Bounds>& bounds)
 void CartPosConstraint::CalcJacobianBlock(const Eigen::Ref<const Eigen::VectorXd>& joint_vals,
                                           Jacobian& jac_block) const
 {
+  auto pose_error_calculator = [&](const Eigen::Ref<const Eigen::VectorXd>& x) {
+    Eigen::Isometry3d new_pose = kinematic_info_->manip->calcFwdKin(x, kinematic_info_->kin_link->link_name);
+    new_pose = kinematic_info_->world_to_base * new_pose * kinematic_info_->kin_link->transform * kinematic_info_->tcp;
+    return target_pose_inv_ * new_pose;
+  };
+
+  // Reserve enough room in the sparse matrix
+  jac_block.reserve(n_dof_ * 6);
+
   if (use_numeric_differentiation)
   {
-    auto error_calculator = [&](const Eigen::Ref<const Eigen::VectorXd>& x) { return this->CalcValues(x); };
-    Jacobian jac0(6, n_dof_);
-    jac0 = calcForwardNumJac(error_calculator, joint_vals, 1e-5);
+    auto error_calculator = [&](const Eigen::Ref<const Eigen::VectorXd>& x) {
+      Eigen::Isometry3d pose_err = pose_error_calculator(x);
+      Eigen::VectorXd err = concat(pose_err.translation(), calcRotationalError2(pose_err.rotation()));
```

Comment:
Why did you remove the other RotationalError calculator and the logic for switching between them?

Reply:
After writing the unit tests, the calcRotationalError breaks down when calculating the jacobian at PI or -PI, but the calcRotationalError2 did not break down at any of the rotation angles between -5PI and 5PI when calculating the jacobian. This is because when using Eigen's angle axis it converts the angle to be between [0, PI] where internally if the angle is between [-PI, 0] it flips the sign of the axis. The calcRotationalError both check for this now flip and reverts it and since the angle is always between [-PI, PI] for Eigen Angle axis switching the range to [0, 2PI] will never be close to 2PI. Though I would expect this function to break down near zero it does not appear to but should investigate further.

Reply:
The reason calcRotationalError2 does not break down at 0 is that we are making sure the angle axis aligns with the quaternion axis which prevents this function from breaking down at zero. 

Angle: 0.001 results in an axis: [0, 0, 1]
Angle: -0.001 results in and axis: [0, 0, -1]
e1 = angle * axis = [0, 0, 0.001]
e2 = angle * axis = [0, 0, -0.001]
delta = e2 - e1 = [0, 0, 0.002]

Reply:
Add this information to the documentation.